change(notification_client): use cross process locking event and media stores instead of in-memory ones

stefanceriu opened the pull request:
This will allow the NSE process to populate the event cache store with even data after retrieving it from sync, as opposed to dropping after rendering the notification it as it did previously.

stefanceriu committed:
change(notification_client): use cross process locking event and media stores instead of in-memory ones

This will allow the NSE process to populate the event cache store with even data after retrieving it from sync, as opposed to dropping after rendering the notification it as it did previously.

Hywan commented on crates/matrix-sdk-base/src/client.rs:
```diff
@@ -221,14 +221,16 @@ impl BaseClient {
+            state_store: BaseStateStore::new(in_memory_store_config.state_store),
+            // The event and media stores both have cross process locking support.
+            event_cache_store: self.event_cache_store.clone(),
+            media_store: self.media_store.clone(),
```
⚠️  This is wrong!

You're cloning the `EventCacheStoreLock` and `MediaStoreLock`, with the main cross-process holder name. It doesn't use the new `cross_process_store_locks_holder_name` received by argument here.

Hywan committed:
fixup

Hywan committed:
refactor(base): Rename `clone_with_in_memory_state_store`.

This patch renames `clone_with_in_memory_state_store` to
`derive_states_for_notification_client`. At least it clarifies what it
does.

Hywan committed:
fixup!

Hywan committed:
fixup!

Hywan commented:
@stefanceriu if you want to give another try, have fun!
